galsim/scene.py

```diff
@@ -194,47 +271,104 @@ def __init__(self, file_name=None, image_dir=None, dir=None, preload=False, nois
+            # zero (generating a RuntimeWarning), and still eliminate those, we will first set that
+            # column to 1.e-5.
+            # We choose a sample-dependent mask ratio cut, since this depends on the peak object
+            # flux, which will differ for the two samples (and we can't really cut on this for
+            # arbitrary user-defined samples).
+            if use_sample == "23.5":
+                cut_ratio = 0.2
+                sn_limit = 20.0
+            else:
+                cut_ratio = 0.8
+                sn_limit = 12.0
+            if self.selection_cat is not None:
+                div_val = self.selection_cat['peak_image_pixel_count']
+                div_val[div_val == 0.] = 1.e-5
+                mask &= ( (self.selection_cat['sn_ellip_gauss'] >= sn_limit) &
+                          ((self.selection_cat['min_mask_dist_pixels'] > 11.0) |
+                           (self.selection_cat['average_mask_adjacent_pixel_count'] / \
+                               div_val < cut_ratio)) )
+
+        if exclusion_level == 'bad_fits' or exclusion_level == 'marginal':
+            # This 'exclusion_level' involves eliminating failed parametric fits (bad fit status
+            # flags).  In this case we only get rid of those with failed bulge+disk AND failed
+            # Sersic fits, so there is no viable parametric model for the galaxy.
             sersicfit_status = self.param_cat['fit_status'][:,4]
             bulgefit_status = self.param_cat['fit_status'][:,0]
-            mask &= ( (sersicfit_status > 0) &
-                      (sersicfit_status < 5) &
-                      (bulgefit_status > 0) &
-                      (bulgefit_status < 5) )
-
-        if exclude_bad:
+            mask &= ( ((sersicfit_status > 0) &
+                      (sersicfit_status < 5)) |
+                      ((bulgefit_status > 0) &
+                      (bulgefit_status < 5)) )
+
+        if exclusion_level == 'marginal':
+            # We have already placed some cuts (above) in this case, but we'll do some more.  For
+            # example, a failed bulge+disk fit often indicates difficulty in fit convergence due to
+            # noisy surface brightness profiles, so we might want to toss out those that have a
+            # failure in EITHER fit.
+            mask &= ( ((sersicfit_status > 0) &
+                      (sersicfit_status < 5)) &
+                      ((bulgefit_status > 0) &
+                      (bulgefit_status < 5)) )
+        
+            # Some fit parameters can indicate a likely sky subtraction error
             hlr = self.param_cat['sersicfit'][:,1]
             n = self.param_cat['sersicfit'][:,2]
-            mask &= ( (n < 5) | (hlr < 1./cosmos_pix_scale) ) 
-            # May add more cuts here if we discover other kinds of problematic objects.
+            mask &= ( (n < 5) | (hlr < 1./cosmos_pix_scale) )
```

Comment:
I think this is saying that the HLR should be smaller than 1 arcsec.  This would be clearer with something like
```
hlr_pix = self.param_cat['sersicfit'][:, 1]
max_hlr_arcsec = 1.0
mask &= ((n < 5) | (hlr_pix < max_hlr_arcsec/cosmos_pix_scale))
```